### PR TITLE
[server][controller][dvc][cc][vpj][producer] Introduce PubSubPositionTypeRegistry

### PR DESCRIPTION
**Pull request by sushantmane**
## Introduce PubSubPositionTypeRegistry and simplify position type handling

This commit introduces `PubSubPositionTypeRegistry`, a configurable and immutable registry  
that manages bidirectional mappings between PubSub position type IDs and their corresponding  
fully qualified factory class names. This registry replaces the older `PubSubPositionType` enum  
and consolidates type resolution logic into one place, enabling compact serialization and  
deserialization of `PubSubPosition` instances. It also validates reserved symbolic types such  
as `EARLIEST`, `LATEST`, and `ApacheKafkaOffsetPosition` to prevent configuration conflicts.  

A new configuration key `pubsub.type.id.to.position.class.name.map` is added to support  
external mappings via `VeniceProperties`. Additional `pubsub.`-prefixed keys are introduced  
for producer, consumer, and admin adapter factories, deprecating the older `pub.sub.` keys  
for consistency. VPJ is updated to propagate all prefixed `pubsub` configs to underlying  
adapters.  

The initialization of `VeniceWriterFactory` is refactored for better modularity and reuse.  
Instead of creating new writer factories in multiple places, a single instance is owned by  
`KafkaStoreIngestionService` and reused across `VeniceViewWriter`, `HelixParticipationService`,  
and `PushStatusStoreWriter`. This improves consistency and avoids redundant setup logic.  

The `PubSubProducerAdapterFactory` and `PubSubConsumerAdapterFactory` interfaces are converted  
to abstract classes with explicit constructors to ensure safe reflective instantiation. Legacy  
factory creation logic and unused fallback methods are removed to simplify the code.  

>  ## AI Generated Summary of Changes
> This pull request introduces several changes to the `da-vinci-client` module, focusing on refactoring the `DaVinciBackend` class, enhancing support for PubSub position deserialization, and improving changelog consumer configurations. The key updates include consolidating PubSub-related logic, introducing a `PubSubPositionDeserializer` for better flexibility, and restructuring the `DaVinciBackend` initialization process.
> 
> ### DaVinciBackend Refactoring:
> * Removed unused imports and variables, such as `VeniceWriterFactory` and `VeniceProperties`, to clean up the `DaVinciBackend` class. [[1]](diffhunk://#diff-929a6b45c574372b5104054afe9a026d7a9f0213d50a573bb669b88a53510d6aL59) [[2]](diffhunk://#diff-929a6b45c574372b5104054afe9a026d7a9f0213d50a573bb669b88a53510d6aL74-L76) [[3]](diffhunk://#diff-929a6b45c574372b5104054afe9a026d7a9f0213d50a573bb669b88a53510d6aL155-L156)
> * Refactored the initialization of `PushStatusStoreWriter` by moving it to the ingestion service and leveraging `VeniceWriterFactory` from the ingestion service instead of creating a new instance. [[1]](diffhunk://#diff-929a6b45c574372b5104054afe9a026d7a9f0213d50a573bb669b88a53510d6aL216-L233) [[2]](diffhunk://#diff-929a6b45c574372b5104054afe9a026d7a9f0213d50a573bb669b88a53510d6aL288-R289)
> 
> ### PubSub Position Deserialization Enhancements:
> * Added a `PubSubPositionDeserializer` to `ChangelogClientConfig` for customizable deserialization of PubSub positions. [[1]](diffhunk://#diff-cec5b773881b7ef815bdca1264c5894b6dd35cd3bfede1eb4a68a78d1bb6ed3cR6) [[2]](diffhunk://#diff-cec5b773881b7ef815bdca1264c5894b6dd35cd3bfede1eb4a68a78d1bb6ed3cR55-R56) [[3]](diffhunk://#diff-cec5b773881b7ef815bdca1264c5894b6dd35cd3bfede1eb4a68a78d1bb6ed3cR121-R124)
> * Updated changelog consumer classes (`InternalLocalBootstrappingVeniceChangelogConsumer`, `LocalBootstrappingVeniceChangelogConsumer`, and `VeniceAfterImageConsumerImpl`) to use the `PubSubPositionDeserializer`. [[1]](diffhunk://#diff-ec3b7dbc9ac5840597a40d9015244b5aaa31566b75c125c7000ad704b2dea923L36-R37) [[2]](diffhunk://#diff-ec3b7dbc9ac5840597a40d9015244b5aaa31566b75c125c7000ad704b2dea923R93-R95) [[3]](diffhunk://#diff-d4b55ac94ea557f4f30f4dc397598ad15a7d51cfa36305a235051d32a0d33f10R3) [[4]](diffhunk://#diff-9b8671d68eab23391851382672c5e3b1fa28d8d46065aa12171c7c73d22290f7R4-R11) [[5]](diffhunk://#diff-9b8671d68eab23391851382672c5e3b1fa28d8d46065aa12171c7c73d22290f7L37-R62)
> 
> ### Changelog Consumer Improvements:
> * Enhanced `ChangelogClientConfig` to include a method for cloning configurations with the `PubSubPositionDeserializer`.
> * Adjusted `VeniceChangeCoordinate` to support decoding with the new deserializer. [[1]](diffhunk://#diff-ec3b7dbc9ac5840597a40d9015244b5aaa31566b75c125c7000ad704b2dea923R158) [[2]](diffhunk://#diff-ec3b7dbc9ac5840597a40d9015244b5aaa31566b75c125c7000ad704b2dea923L462-R466) [[3]](diffhunk://#diff-b613912deb10ce3801a28541ad821c1a2df9f9bae0bb3b8f2675863788df86e8R5)
> 
> These changes streamline PubSub integration and improve the maintainability of the `da-vinci-client` module.
> 

###  Code changes
- [x] Added new code behind **a config**. If so list the config names and their default values in the PR description.
- [x] Introduced new **log lines**. 
  - [x] Confirmed if logs need to be **rate limited** to avoid excessive logging.

###  **Concurrency-Specific Checks**
Both reviewer and PR author to verify
- [x] Code has **no race conditions** or **thread safety issues**.
- [x] Proper **synchronization mechanisms** (e.g., `synchronized`, `RWLock`) are used where needed.
- [x] No **blocking calls** inside critical sections that could lead to deadlocks or performance degradation.
- [x] Verified **thread-safe collections** are used (e.g., `ConcurrentHashMap`, `CopyOnWriteArrayList`).
- [x] Validated proper exception handling in multi-threaded code to avoid silent thread termination.

## How was this PR tested?

- [x] New unit tests added.
- [ ] New integration tests added.
- [x] Modified or extended existing tests.
- [x] Verified backward compatibility (if applicable).

## Does this PR introduce any user-facing or breaking changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Clearly explain the behavior change and its impact.

### CODE REVIEW

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+  public static final Int2ObjectMap<String> RESERVED_POSITION_TYPE_ID_TO_CLASS_NAME_MAP;
+
+  static {
+    Int2ObjectMap<String> tempMap = new Int2ObjectOpenHashMap<>(3);
+    tempMap.put(EARLIEST_POSITION_RESERVED_TYPE_ID, EarliestPositionFactory.class.getName());
+    tempMap.put(LATEST_POSITION_RESERVED_TYPE_ID, LatestPositionFactory.class.getName());
+    tempMap.put(APACHE_KAFKA_OFFSET_POSITION_TYPE_ID, ApacheKafkaOffsetPositionFactory.class.getName());
+
+    Int2ObjectMap<PubSubPositionFactory> factoryMap = new Int2ObjectOpenHashMap<>(3);
+    factoryMap.put(EARLIEST_POSITION_RESERVED_TYPE_ID, new EarliestPositionFactory(EARLIEST_POSITION_RESERVED_TYPE_ID));
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java:85]: Just wondering...Why can't we just refer EARLIEST_POSITION_RESERVED_TYPE_ID in EarliestPositionFactory and same for Latest?
This param passing looks unnecessary as we should not have other type ID.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+      throw new VeniceException("PubSub position type ID not found: " + typeId);
+    }
+    return typeIdToFactoryMap.get(typeId);
+  }
+
+  /**
+   * Returns the mapping of integer type IDs to fully qualified factory class names.
+   * @return a map from integer type IDs to the fully qualified class factory names of {@code PubSubPosition} implementations
+   */
+  @VisibleForTesting
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java:269]: nit: move the non-public method to the bottom.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubUtil.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      brokerAddress = properties.getProperty(KAFKA_BOOTSTRAP_SERVERS);
+    }
+    return brokerAddress;
+  }
+
+  public static String getPubSubBrokerAddress(VeniceProperties properties) {
+    return properties.getStringWithAlternative(PUBSUB_BROKER_ADDRESS, KAFKA_BOOTSTRAP_SERVERS);
+  }
+
+  public static String getPubSubBrokerAddress(VeniceProperties properties, String defaultValue) {
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubUtil.java:25]: nit: rename to withDefault?

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientDiskFullTest.java [hunk 4]
```diff
@@ -207,6 +208,8 @@ public void testDaVinciDiskFullFailure(boolean useDaVinciSpecificExecutionStatus
     String inputDirPath = "file://" + inputDir.getAbsolutePath();
     Schema recordSchema = TestWriteUtils.writeSimpleAvroFileWithStringToStringSchema(inputDir, 1, 10);
     Properties vpjProperties = defaultVPJProps(venice, inputDirPath, storeName);
+    vpjProperties.putAll(
+        PubSubBrokerWrapper.getBrokerDetailsForClients(Collections.singletonList(venice.getPubSubBrokerWrapper())));
```
Comment (sixpluszero) [inline internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientDiskFullTest.java:212]: Should it be merged into defaultVPJProps? Also I am a bit confused could you explain which part is using this?

File: internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  public void testHashCodeConsistency() {
+    LatestPosition instance1 = LatestPosition.getInstance();
+    LatestPosition instance2 = LatestPosition.getInstance();
+    assertEquals(instance1.hashCode(), instance2.hashCode(), "Hash codes should match");
+  }
+
+  @Test
+  public void testNumericOffset() {
+    LatestPosition instance = LatestPosition.getInstance();
+    assertEquals(instance.getNumericOffset(), -1L, "Numeric offset should be -1");
```
Comment (sixpluszero) [inline internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java:66]: I have a question though - so for the Earliest and Latest Position, we are not supposed to use numericOffset then?

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicManager.java [hunk 6]
```diff
@@ -801,6 +802,18 @@ public String getPubSubClusterAddress() {
     return this.pubSubClusterAddress;
   }
 
+  /**
+   * Resolve the position bytes to a PubSubPosition object.
+   * @param topicPartition the topic partition to resolve the position for
+   * @param positionBytes the position bytes to resolve
+   * @return the resolved PubSubPosition object
+   */
+  public PubSubPosition resolvePosition(PubSubTopicPartition topicPartition, byte[] positionBytes) {
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicManager.java:811]: Is this not used anywhere? This null looks fishy If not let's remove it for now and add back when it is ready?

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubUtil.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import static com.linkedin.venice.ConfigKeys.KAFKA_BOOTSTRAP_SERVERS;
+import static com.linkedin.venice.ConfigKeys.PUBSUB_BROKER_ADDRESS;
+import static com.linkedin.venice.pubsub.PubSubConstants.PUBSUB_CLIENT_CONFIG_PREFIX;
+
+import com.linkedin.venice.utils.Utils;
+import com.linkedin.venice.utils.VeniceProperties;
+import java.util.Properties;
+
+
+public final class PubSubUtil {
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubUtil.java:12]: nice to have this uniformed naming util for better debugging purpose.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/producer/ApacheKafkaProducerConfig.java [hunk 8]
```diff
@@ -253,7 +238,45 @@ public static void copyKafkaSASLProperties(Properties configuration, Properties
...
+   * mapping is consistent with the reserved value. This helps detect misconfigured or corrupted mappings.
+   * <p>
+   * If the position class is not registered or the associated type ID does not match the expected reserved value,
+   * this method logs a descriptive error and throws a {@link VeniceException}.
+   *
+   * @param typeRegistry the position type registry to validate
+   * @throws VeniceException if the type is missing or mismatched
+   */
+  private static void validateKafkaPositionType(PubSubPositionTypeRegistry typeRegistry) {
+    if (!typeRegistry.containsFactoryClass(KAFKA_POSITION_FACTORY_CLASS_NAME)) {
```
Comment (sixpluszero) [inline internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/producer/ApacheKafkaProducerConfig.java:260]: Meaning we will always do strict config check? Maybe it is good, but I wonder do you know when we set strict check to false in the past?

**Review by sixpluszero — COMMENTED** [review]
Thank you @sushantmane for the change! I finished a pass and left some comments (mostly questions because I don't have full context on some part)

File: internal/venice-common/src/main/java/com/linkedin/venice/offsets/OffsetRecord.java [hunk 9]
```diff
@@ -316,15 +316,6 @@ public void setLatestConsumedVtOffset(long latestConsumedVtOffset) {
         ApacheKafkaOffsetPosition.of(latestConsumedVtOffset).getPositionWireFormat().getRawBytes());
   }
 
-  public long getLatestConsumedVtOffset() {
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/offsets/OffsetRecord.java:319]: I see. We need to fix the impl of this method though. It's incorrect right now.

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubClientsFactory.java [hunk 10]
```diff
@@ -93,19 +100,21 @@ public static PubSubAdminAdapterFactory<PubSubAdminAdapter> createSourceOfTruthA
...
+      String preferredConfigKey,
+      String alternateConfigKey,
       String defaultClassName,
       FactoryType factoryType) {
     String className;
-    if (properties.containsKey(configKey)) {
-      className = properties.getString(configKey);
+    if (properties.containsKey(preferredConfigKey) || properties.containsKey(alternateConfigKey)) {
+      className = properties.getStringWithAlternative(preferredConfigKey, alternateConfigKey);
       LOGGER.debug("Creating pub-sub {} adapter factory instance for class: {}", factoryType, className);
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubClientsFactory.java:118]: Currently, this object is created multiple times because we recreate the VeniceServerConfig each time a PubSub consumer is requested for subscription 🤦 . In my upcoming PR, I’ll re-enable this log line as part of cleaning up the PubSub client configurations and preventing unnecessary recreation of the VeniceServerConfig object.

File: internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientDiskFullTest.java (hunk 4 above)
Comment (sushantmane) [reply internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/endToEnd/DaVinciClientDiskFullTest.java:212]: Yeah, we’re planning to replace it eventually. But for now, I’m skipping it because doing so
would require updating all the tests, which I’d prefer to defer.

> Also I am a bit confused could you explain which part is using this?


There are some PubSub-specific properties that need to be passed down to the PubSub adapters.
Depending on the type of PubSub backend, these properties can vary. Each PubSub backend
needs to override the client properties API to supply the appropriate custom properties,
which ensures the corresponding clients are created correctly.

File: internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java (hunk 5 above)
Comment (sushantmane) [reply internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java:66]: Ideally not. Eventually there won't be any  `getNumericOffset` api for any position type.

Comment (sixpluszero) [reply internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java:66]: Oh if this is the case should we throw instead? Or do you think there is some other edge case?

File: internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java (hunk 1 above)
Comment (sixpluszero) [reply internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionTypeRegistry.java]: got it..sure let's keep it as what it is

**Review by sixpluszero — APPROVED** [review]
LGTM. Synced offline to get resolved my questions. Thank you for your change.

File: internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java (hunk 5 above)
Comment (sushantmane) [reply internal/venice-common/src/test/java/com/linkedin/venice/pubsub/api/LatestPositionTest.java:66]: Updated to throw an exception

**Comment by sushantmane** [comment]
Thanks a lot, @sixpluszero for the review! 🙏

**Review by sixpluszero — APPROVED** [review]
